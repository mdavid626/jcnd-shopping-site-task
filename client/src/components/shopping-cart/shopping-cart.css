.ShoppingCart {
  background-color: #f2f2f2;
  border-radius: 5px;
  padding: var(--space3);
  display: flex;
  flex-direction: column;
  gap: var(--space3);
  min-width: 700px;
  min-height: 400px;
}

.ShoppingCart-header {
  font-size: 22px;
}

.ShoppingCart-content {
  display: flex;
  flex-direction: column;
  flex: 1;
}

.ShoppingCart-close {
  cursor: pointer;
  color: #b3005e;
}

.ShoppingCart-items {
  display: flex;
  flex-direction: column;
  gap: var(--space3);
  overflow-y: scroll;
  max-height: 280px;
  padding: 0 var(--space2) 0 0;
  flex: 1;
}

.ShoppingCart-total {
  align-self: flex-end;
  font-size: 18px;
}

.ShoppingCart-actions {
  display: flex;
  gap: var(--space3);
  align-items: center;
  align-self: flex-end;
}

.ShoppingCart-placeOrder {
  cursor: pointer;
  font-size: 18px;
  background-color: var(--primary-color);
  padding: var(--space2) var(--space3);
  color: #f2f2f2;
  border-radius: 3px;
}
