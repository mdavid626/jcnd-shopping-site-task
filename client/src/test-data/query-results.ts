export const productsQueryResult = {
  data: {
    products: {
      nodes: [
        {
          id: '63ea6485732ff5cf8b364431',
          priceInCents: 9699,
          name: 'Broccoli',
          inStock: 466,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364432',
          priceInCents: 3434,
          name: 'Broccolini',
          inStock: 223,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364433',
          priceInCents: 3692,
          name: 'Brussels Sprouts',
          inStock: 12,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364434',
          priceInCents: 8839,
          name: 'Cabbage',
          inStock: 200,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364435',
          priceInCents: 6828,
          name: 'Carrots',
          inStock: 151,
          __typename: 'Product',
        },
      ],
      pageInfo: { page: 1, numberOfPages: 15, __typename: 'PageInfo' },
      __typename: 'ProductsQueryResult',
    },
  },
};

export const placeOrderQueryResult1 = {
  data: { placeOrder: '63ede5e031f3fc4012a0951b' },
};
