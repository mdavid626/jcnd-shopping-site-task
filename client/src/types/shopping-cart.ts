export type ShoppingCartItem = {
  productId: string;
  amount: number;
};
