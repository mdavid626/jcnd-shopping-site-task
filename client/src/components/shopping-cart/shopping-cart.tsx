import React from 'react';
import {
  usePlaceOrder,
  useShoppingCart,
} from '../../hooks/shopping-car-thooks/shopping-cart-hooks';
import { ShoppingCartItem } from '../../types/shopping-cart';
import Currency from '../currency/currency';
import ShoppingCartListItem from '../shopping-cart-list-item/shopping-cart-list-item';
import './shopping-cart.css';

const calculateTotal = (shoppingCart: ShoppingCartItem[]) =>
  shoppingCart.reduce(
    (acc, item) => acc + item.product.priceInCents * item.amount,
    0
  );

const ShoppingCart: React.FC<{ onClose: () => void }> = ({ onClose }) => {
  const [shoppingCart] = useShoppingCart();
  const [placeOrder, isPlacingOrder] = usePlaceOrder();
  const handlePlaceOrderClick = async () => {
    await placeOrder();
    onClose();
  };
  return (
    <div className="ShoppingCart">
      <div className="ShoppingCart-header">Shopping Cart</div>
      <div className="ShoppingCart-content">
        {shoppingCart.length > 0 ? (
          <div className="ShoppingCart-items">
            {shoppingCart.map((item) => (
              <ShoppingCartListItem
                key={item.product.id}
                shoppingCartItem={item}
              />
            ))}
          </div>
        ) : (
          <div>Your shopping cart is empty</div>
        )}
      </div>
      {shoppingCart.length > 0 && (
        <div className="ShoppingCart-total" data-testid="ShoppingCart-total">
          Total: <Currency priceInCents={calculateTotal(shoppingCart)} />
        </div>
      )}
      <div className="ShoppingCart-actions">
        <div onClick={() => onClose()} className="ShoppingCart-close">
          close
        </div>
        {shoppingCart.length > 0 && !isPlacingOrder && (
          <div
            onClick={handlePlaceOrderClick}
            className="ShoppingCart-placeOrder"
          >
            Place order
          </div>
        )}
        {isPlacingOrder && (
          <div className="ShoppingCart-placeOrder">Placing order...</div>
        )}
      </div>
    </div>
  );
};

export default ShoppingCart;
