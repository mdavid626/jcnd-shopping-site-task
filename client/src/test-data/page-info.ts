import { PageInfo } from '../types/page-info';

export const testPageInfo1: PageInfo = {
  page: 1,
  numberOfPages: 5,
};
