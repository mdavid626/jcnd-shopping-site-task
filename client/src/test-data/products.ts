import { Product, ProductsQueryResult } from '../types/products';

export const testProduct1: Product = {
  id: '63ea6485732ff5cf8b36442c',
  name: 'Artichoke',
  priceInCents: 4636,
  inStock: 5,
};

export const testProduct2: Product = {
  id: '63ea6485732ff5cf8b36442d',
  name: 'Asparagus',
  priceInCents: 2536,
  inStock: 3,
};

export const testProductQueryResult1: ProductsQueryResult = {
  products: {
    nodes: [testProduct1, testProduct2],
    pageInfo: {
      page: 1,
      numberOfPages: 2,
    },
  },
};
