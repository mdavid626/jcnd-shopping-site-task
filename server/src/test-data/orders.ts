import { OrderRequestItem } from '../types/order';

export const testOrderRequestItems1: OrderRequestItem[] = [
  { productId: '63ea6485732ff5cf8b36442c', amount: 2 },
  { productId: '63ea6485732ff5cf8b36442d', amount: 4 },
  { productId: '63ea6485732ff5cf8b36442e', amount: 6 },
];
