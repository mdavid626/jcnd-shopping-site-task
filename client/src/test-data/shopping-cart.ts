import { ShoppingCartItem } from '../types/shopping-cart';
import { testProduct1, testProduct2 } from './products';

export const testShoppingCartItem1: ShoppingCartItem = {
  product: testProduct1,
  amount: 3,
};

export const testShoppingCartItem2: ShoppingCartItem = {
  product: testProduct2,
  amount: 36,
};

export const testShoppingCart1: ShoppingCartItem[] = [
  testShoppingCartItem1,
  testShoppingCartItem2,
];
