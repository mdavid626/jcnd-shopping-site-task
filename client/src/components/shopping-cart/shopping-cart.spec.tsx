import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import {
  usePlaceOrder,
  useShoppingCart,
} from '../../hooks/shopping-car-thooks/shopping-cart-hooks';
import { testShoppingCart1 } from '../../test-data/shopping-cart';
import ShoppingCartListItem from '../shopping-cart-list-item/shopping-cart-list-item';
import ShoppingCart from './shopping-cart';

jest.mock('../../hooks/shopping-car-thooks/shopping-cart-hooks');
jest.mock('../shopping-cart-list-item/shopping-cart-list-item');

describe('shopping-cart', () => {
  beforeEach(() => {
    (ShoppingCartListItem as jest.Mock).mockReturnValue(
      <div>shopping-cart-list-item</div>
    );
    (useShoppingCart as jest.Mock).mockReturnValue([
      testShoppingCart1,
      jest.fn(),
    ]);
    (usePlaceOrder as jest.Mock).mockReturnValue([jest.fn(), false]);
  });
  afterEach(cleanup);
  afterEach(jest.resetAllMocks);

  it('should render', () => {
    const { asFragment } = render(<ShoppingCart onClose={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
    expect(useShoppingCart).toHaveBeenCalledWith();
    expect(usePlaceOrder).toHaveBeenCalledWith();
    expect(ShoppingCartListItem).toHaveBeenCalledTimes(2);
    expect(ShoppingCartListItem).toHaveBeenCalledWith(
      { shoppingCartItem: testShoppingCart1[0] },
      {}
    );
    expect(ShoppingCartListItem).toHaveBeenCalledWith(
      { shoppingCartItem: testShoppingCart1[1] },
      {}
    );
  });

  it('should render message when no items in shopping cart', () => {
    (useShoppingCart as jest.Mock).mockReturnValue([[], jest.fn()]);
    render(<ShoppingCart onClose={jest.fn()} />);
    expect(screen.getByText('Your shopping cart is empty')).toBeVisible();
    expect(screen.queryByTestId('ShoppingCart-total')).toBe(null);
    expect(screen.queryByText('Place order')).toBe(null);
    expect(ShoppingCartListItem).not.toHaveBeenCalled();
  });

  it('should close when close button clicked', async () => {
    const handleClose = jest.fn();
    render(<ShoppingCart onClose={handleClose} />);
    await userEvent.click(screen.getByText('close'));
    expect(handleClose).toHaveBeenCalledWith();
  });

  it('should place order when place order button clicked', async () => {
    const placeOrder = jest.fn().mockResolvedValue(undefined);
    (usePlaceOrder as jest.Mock).mockReturnValue([placeOrder, false]);
    const handleClose = jest.fn();
    render(<ShoppingCart onClose={handleClose} />);
    await userEvent.click(screen.getByText('Place order'));
    expect(placeOrder).toHaveBeenCalledWith();
    expect(handleClose).toHaveBeenCalledWith();
  });

  it('should show loading text when placing order', async () => {
    (usePlaceOrder as jest.Mock).mockReturnValue([jest.fn(), true]);
    render(<ShoppingCart onClose={jest.fn()} />);
    expect(screen.getByText('Placing order...')).toBeVisible();
    expect(screen.queryByText('Place order')).toBe(null);
  });
});
