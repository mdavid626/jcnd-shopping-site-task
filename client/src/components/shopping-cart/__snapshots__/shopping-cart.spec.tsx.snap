// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shopping-cart should render 1`] = `
<DocumentFragment>
  <div
    class="ShoppingCart"
  >
    <div
      class="ShoppingCart-header"
    >
      Shopping Cart
    </div>
    <div
      class="ShoppingCart-content"
    >
      <div
        class="ShoppingCart-items"
      >
        <div>
          shopping-cart-list-item
        </div>
        <div>
          shopping-cart-list-item
        </div>
      </div>
    </div>
    <div
      class="ShoppingCart-total"
      data-testid="ShoppingCart-total"
    >
      Total: 1.052,04 €
    </div>
    <div
      class="ShoppingCart-actions"
    >
      <div
        class="ShoppingCart-close"
      >
        close
      </div>
      <div
        class="ShoppingCart-placeOrder"
      >
        Place order
      </div>
    </div>
  </div>
</DocumentFragment>
`;
